src/Sulu/Bundle/CategoryBundle/Controller/KeywordController.php

```diff
@@ -0,0 +1,317 @@
+        $listBuilder->distinct(true);
+        $listBuilder->addGroupBy($fieldDescriptor['id']);
+
+        $listResponse = $listBuilder->execute();
+
+        $list = new ListRepresentation(
+            $listResponse,
+            self::$entityKey,
+            'get_category_keywords',
+            array_merge(['categoryId' => $categoryId], $request->query->all()),
+            $listBuilder->getCurrentPage(),
+            $listBuilder->getLimit(),
+            $listBuilder->count()
+        );
+
+        return $this->handleView($this->view($list, 200));
+    }
+
+    /**
+     * Creates new keyword for given category.
+     *
+     * @param int $categoryId
+     * @param Request $request
+     *
+     * @return Response
+     */
+    public function postAction($categoryId, Request $request)
+    {
+        /** @var Keyword $keyword */
+        $keyword = $this->getKeywordRepository()->createNew();
+        $category = $this->getCategoryManager()->findById($categoryId);
+        $keyword->setKeyword($request->get('keyword'));
+        $keyword->setLocale($request->get('locale'));
+
+        $keyword = $this->getKeywordManager()->save($keyword, $category);
+
+        $this->getEntityManager()->persist($keyword);
+        $this->getEntityManager()->flush();
+
+        return $this->handleView($this->view($keyword));
+    }
+
+    /**
+     * Updates given keyword for given category.
+     *
+     * @param int $categoryId
+     * @param int $keywordId
+     * @param Request $request
+     *
+     * @return Response
+     *
+     * @throws KeywordIsMultipleReferencedException
+     * @throws KeywordNotUniqueException
+     */
+    public function putAction($categoryId, $keywordId, Request $request)
+    {
+        $keyword = $this->getKeywordRepository()->findById($keywordId);
+
+        if (!$keyword) {
+            return $this->handleView($this->view(null, 404));
+        }
+
+        $force = $request->get('force');
+
+        // overwrite existing keyword if force is present
+        if (null === $force
+            && $keyword->isReferencedMultiple()
+            && in_array($force, [self::FORCE_OVERWRITE, self::FORCE_DETACH, self::FORCE_MERGE, null])
+        ) {
+            // return conflict if keyword is used by other categories
+            throw new KeywordIsMultipleReferencedException($keyword);
+        }
+
+        $category = $this->getCategoryManager()->findById($categoryId);
+
+        if ($force !== self::FORCE_MERGE
+            && $this->getKeywordRepository()->findByKeyword($request->get('keyword'), $keyword->getLocale()) !== null
+        ) {
+            throw new KeywordNotUniqueException($keyword);
+        }
+
+        if ($force === self::FORCE_DETACH || $force === self::FORCE_MERGE) {
+            $keyword = $this->handleDetach($category, $keyword, $request->get('keyword'));
+        } else {
+            $keyword = $this->handleOverwrite($category, $keyword, $request->get('keyword'));
+        }
```

Comment:
The discussion about this got lost during the last commits... IMO what we see here is part of the business logic (throwing these exceptions and merging, detaching or overriding the keywords), so I would say this logic should go into the manager. The flag from the request could still be read here, and passed as a single boolean value to the manager.

@webmozart @chirimoya @dantleech WDYT? Is a critical decision regarding our future architecture, that's why I would like to also have your opinions on that :smiley: